include/core/mir/mutex.h

```diff
@@ -48,14 +55,28 @@ class MutexGuard
 
     Guarded& operator*()
     {
-        return value;
+        return *value;
     }
+
     Guarded* operator->()
     {
-        return &value;
+        return value;
+    }
+
+    /**
+     * Relinquish access to the mutex-guarded data
+     *
+     * This prevents further access to the contained data and unlocks
+     * the associated mutex.
+     */
+    void drop()
+    {
+        value = nullptr;
+        lock.unlock();
     }
+
 private:
-    Guarded& value;
+    Guarded* value;
     std::unique_lock<std::mutex> lock;
```

Comment:
It is unfortunate that we need to use `std::unique_lock` and can't optimise with `std::lock_guard` where that provides the required guarantees.